File: app/src/main/java/com/mecong/tenderalarm/model/AlarmEntity.java
package com.mecong.tenderalarm.model;

import android.database.Cursor;

import com.mecong.tenderalarm.R;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import static java.util.concurrent.TimeUnit.MINUTES;


public class AlarmEntity {
    static final int
            MO_BINARY = 0b10000000, TU_BINARY = 0b01000000, WE_BINARY = 0b00100000,
            TH_BINARY = 0b00010000,
            SA_BINARY = 0b00000100, FR_BINARY = 0b00001000, SU_BINARY = 0b00000010;
    long id;
    int hour;
    int minute;

    int days = 0;

    long exactDate = 0;

    int complexity = 1;

    String message;

    boolean active = true;
    int ticksTime;

    String melodyUrl;

    String melodyName;

    String vibrationType;

    Integer volume;

    Integer snoozeMaxTimes = 10;

    Integer canceledNextAlarms = 0;

    Long nextTime = -1L;

    Long nextNotCanceledTime = -1L;

    Integer nextRequestCode = -1;
    boolean timeToSleepNotification;
    boolean headsUp;

    public AlarmEntity() {
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public int getHour() {
        return hour;
    }

    public void setHour(int hour) {
        this.hour = hour;
    }

    public int getMinute() {
        return minute;
    }

    public void setMinute(int minute) {
        this.minute = minute;
    }

    public int getDays() {
        return days;
    }

    public void setDays(int days) {
        this.days = days;
    }

    public long getExactDate() {
        return exactDate;
    }

    public void setExactDate(long exactDate) {
        this.exactDate = exactDate;
    }

    public int getComplexity() {
        return complexity;
    }

    public void setComplexity(int complexity) {
        this.complexity = complexity;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public boolean isActive() {
        return active;
    }

    public void setActive(boolean active) {
        this.active = active;
    }

    public int getTicksTime() {
        return ticksTime;
    }

    public void setTicksTime(int ticksTime) {
        this.ticksTime = ticksTime;
    }

    public String getMelodyUrl() {
        return melodyUrl;
    }

    public void setMelodyUrl(String melodyUrl) {
        this.melodyUrl = melodyUrl;
    }

    public String getMelodyName() {
        return melodyName;
    }

    public void setMelodyName(String melodyName) {
        this.melodyName = melodyName;
    }

    public String getVibrationType() {
        return vibrationType;
    }

    public void setVibrationType(String vibrationType) {
        this.vibrationType = vibrationType;
    }

    public Integer getVolume() {
        return volume;
    }

    public void setVolume(Integer volume) {
        this.volume = volume;
    }

    public Integer getSnoozeMaxTimes() {
        return snoozeMaxTimes;
    }

    public void setSnoozeMaxTimes(Integer snoozeMaxTimes) {
        this.snoozeMaxTimes = snoozeMaxTimes;
    }

    public Integer getCanceledNextAlarms() {
        return canceledNextAlarms;
    }

    public void setCanceledNextAlarms(Integer canceledNextAlarms) {
        this.canceledNextAlarms = canceledNextAlarms;
    }

    public Long getNextNotCanceledTime() {
        return nextNotCanceledTime;
    }

    public void setNextNotCanceledTime(Long nextNotCanceledTime) {
        this.nextNotCanceledTime = nextNotCanceledTime;
    }

    public Integer getNextRequestCode() {
        return nextRequestCode;
    }

    public void setNextRequestCode(Integer nextRequestCode) {
        this.nextRequestCode = nextRequestCode;
    }

    public boolean isTimeToSleepNotification() {
        return timeToSleepNotification;
    }

    public void setTimeToSleepNotification(boolean timeToSleepNotification) {
        this.timeToSleepNotification = timeToSleepNotification;
    }

    public boolean isHeadsUp() {
        return headsUp;
    }

    public void setHeadsUp(boolean headsUp) {
        this.headsUp = headsUp;
    }

    public Long getNextTime() {
        return nextTime;
    }

    public void setNextTime(Long nextTime) {
        this.nextTime = nextTime;
    }

    public AlarmEntity(Cursor cursor) {

        this.id = cursor.getLong(cursor.getColumnIndex("_id"));
        this.hour = cursor.getInt(cursor.getColumnIndex("hour"));
        this.minute = cursor.getInt(cursor.getColumnIndex("minute"));
        this.days = cursor.getInt(cursor.getColumnIndex("days"));
        this.complexity = cursor.getInt(cursor.getColumnIndex("complexity"));
        this.exactDate = cursor.getLong(cursor.getColumnIndex("exact_date"));
        this.message = cursor.getString(cursor.getColumnIndex("message"));
        this.active = cursor.getInt(cursor.getColumnIndex("active")) == 1;
        this.ticksTime = cursor.getInt(cursor.getColumnIndex("ticks_time"));
        this.melodyUrl = cursor.getString(cursor.getColumnIndex("melody_url"));
        this.melodyName = cursor.getString(cursor.getColumnIndex("melody_name"));
        this.vibrationType = cursor.getString(cursor.getColumnIndex("vibration_type"));
        this.volume = cursor.getInt(cursor.getColumnIndex("volume"));
        this.snoozeMaxTimes = cursor.getInt(cursor.getColumnIndex("snooze_max_times"));
        this.canceledNextAlarms = cursor.getInt(cursor.getColumnIndex("canceled_next_alarms"));
        this.nextTime = cursor.getLong(cursor.getColumnIndex("next_time"));
        this.nextNotCanceledTime = cursor.getLong(cursor.getColumnIndex("next_not_canceled_time"));
        this.nextRequestCode = cursor.getInt(cursor.getColumnIndex("next_request_code"));

        this.headsUp = cursor.getInt(cursor.getColumnIndex("heads_up")) == 1;
        this.timeToSleepNotification = cursor.getInt(cursor.getColumnIndex("tts_notification")) == 1;
    }

    public static int daysMarshaling(boolean mo, boolean tu, boolean we, boolean th, boolean fr, boolean sa, boolean su) {
        int d = 0;
        if (mo) {
            d |= MO_BINARY;
        }

        if (tu) {
            d |= TU_BINARY;
        }

        if (we) {
            d |= WE_BINARY;
        }

        if (th) {
            d |= TH_BINARY;
        }

        if (fr) {
            d |= FR_BINARY;
        }

        if (sa) {
            d |= SA_BINARY;
        }

        if (su) {
            d |= SU_BINARY;
        }

        return d;
    }

    public long getNextTimeWithTicks() {
        return getNextTime() + MINUTES.toMillis(getTicksTime());
    }

    //@ToString.Include
    private String nextTime() {
        SimpleDateFormat format = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US);
        return format.format(new Date(nextTime));
    }

    public void updateNextAlarmDate(boolean manually) {
        Calendar calendar = Calendar.getInstance();
        Calendar calendarNow = Calendar.getInstance();
        nextTime = -1L;

        if (exactDate == 0) {
            calendar.set(Calendar.HOUR_OF_DAY, hour);
            calendar.set(Calendar.MINUTE, minute);
            calendar.set(Calendar.SECOND, 0);
            calendar.set(Calendar.MILLISECOND, 0);

            if (!manually) {
                calendarNow.add(Calendar.MINUTE, ticksTime);
            }

            if (calendar.before(calendarNow)) {
                calendar.add(Calendar.DAY_OF_YEAR, 1);
            }

            if (days > 0) {
                List<Boolean> allDaysAsList = allDaysAsList();
                int i = calendar.get(Calendar.DAY_OF_WEEK) - 1;
                int skip = canceledNextAlarms;
                while (true) {
                    if (i >= allDaysAsList.size()) {
                        i = 0;
                    }

                    final boolean fireAtThisDayOfWeek = allDaysAsList.get(i);
                    if (fireAtThisDayOfWeek) {
                        if (skip <= 0) {
                            break;
                        } else {
                            skip--;
                            if (nextTime == -1) {
                                nextTime = calendar.getTimeInMillis();
                            }
                        }
                    }

                    calendar.add(Calendar.DAY_OF_YEAR, 1);
                    i++;
                }
            }
        } else {
            calendar.setTimeInMillis(exactDate);
            if (calendar.before(calendarNow)) {
                calendar.add(Calendar.DAY_OF_YEAR, 1);
                exactDate = calendar.getTimeInMillis();
            }
        }

        nextNotCanceledTime = calendar.getTimeInMillis();
        if (nextTime == -1) {
            if (ticksTime > 0) {
                calendar.add(Calendar.MINUTE, -ticksTime);
            }
            nextTime = calendar.getTimeInMillis();
        }

        this.nextRequestCode = getRequestCode();
    }

    private int getRequestCode() {
        return (int) id * 100000;
    }

    private List<Boolean> allDaysAsList() {
        List<Boolean> allDaysAsList = new ArrayList<>(8);
        allDaysAsList.add((days & SU_BINARY) == SU_BINARY);
        allDaysAsList.add((days & MO_BINARY) == MO_BINARY);
        allDaysAsList.add((days & TU_BINARY) == TU_BINARY);
        allDaysAsList.add((days & WE_BINARY) == WE_BINARY);
        allDaysAsList.add((days & TH_BINARY) == TH_BINARY);
        allDaysAsList.add((days & FR_BINARY) == FR_BINARY);
        allDaysAsList.add((days & SA_BINARY) == SA_BINARY);
        return allDaysAsList;
    }

    public Map<Integer, Boolean> getDaysAsMap() {
        Map<Integer, Boolean> daysMap = new LinkedHashMap<>();
        daysMap.put(R.string.su, (days & SU_BINARY) == SU_BINARY);
        daysMap.put(R.string.mo, (days & MO_BINARY) == MO_BINARY);
        daysMap.put(R.string.tu, (days & TU_BINARY) == TU_BINARY);
        daysMap.put(R.string.we, (days & WE_BINARY) == WE_BINARY);
        daysMap.put(R.string.th, (days & TH_BINARY) == TH_BINARY);
        daysMap.put(R.string.fr, (days & FR_BINARY) == FR_BINARY);
        daysMap.put(R.string.sa, (days & SA_BINARY) == SA_BINARY);
        return daysMap;
    }
}

File: app/src/main/java/com/mecong/tenderalarm/model/AlarmEntity.kt
package com.mecong.tenderalarm.model

import android.database.Cursor
import com.mecong.tenderalarm.R
import java.text.SimpleDateFormat
import java.util.*
import java.util.concurrent.TimeUnit

data class AlarmEntity(var id: Long = -1, var hour: Int = 0, var minute: Int = 0, var days: Int = 0,
                       var exactDate: Long = 0, var ticksTime: Int = 0,
                       var ticksType: Int = 0,
                       var melodyUrl: String? = null, var melodyName: String? = null,
                       var vibrationType: String? = null, var volume: Int? = null,
                       var snoozeMaxTimes: Int = 10, var canceledNextAlarms: Int = 0,
                       var complexity: Int = 1, var message: String? = null,
                       var isActive: Boolean = true, var nextTime: Long = -1L,
                       var nextNotCanceledTime: Long = -1L, var nextRequestCode: Int = -1,
                       var increaseVolume: Int = 5,
                       var isTimeToSleepNotification: Boolean = false, var isHeadsUp: Boolean = false) {


    constructor(cursor: Cursor) : this(
            id = cursor.getLong(cursor.getColumnIndex("_id")),
            hour = cursor.getInt(cursor.getColumnIndex("hour")),
            minute = cursor.getInt(cursor.getColumnIndex("minute")),
            days = cursor.getInt(cursor.getColumnIndex("days")),
            complexity = cursor.getInt(cursor.getColumnIndex("complexity")),
            exactDate = cursor.getLong(cursor.getColumnIndex("exact_date")),
            message = cursor.getString(cursor.getColumnIndex("message")),
            isActive = cursor.getInt(cursor.getColumnIndex("active")) == 1,
            ticksTime = cursor.getInt(cursor.getColumnIndex("ticks_time")),
            ticksType = cursor.getInt(cursor.getColumnIndex("ticks_type")),
            melodyUrl = cursor.getString(cursor.getColumnIndex("melody_url")),
            melodyName = cursor.getString(cursor.getColumnIndex("melody_name")),
            vibrationType = cursor.getString(cursor.getColumnIndex("vibration_type")),
            volume = cursor.getInt(cursor.getColumnIndex("volume")),
            snoozeMaxTimes = cursor.getInt(cursor.getColumnIndex("snooze_max_times")),
            canceledNextAlarms = cursor.getInt(cursor.getColumnIndex("canceled_next_alarms")),
            nextTime = cursor.getLong(cursor.getColumnIndex("next_time")),
            nextNotCanceledTime = cursor.getLong(cursor.getColumnIndex("next_not_canceled_time")),
            nextRequestCode = cursor.getInt(cursor.getColumnIndex("next_request_code")),
            isHeadsUp = cursor.getInt(cursor.getColumnIndex("heads_up")) == 1,
            isTimeToSleepNotification = cursor.getInt(cursor.getColumnIndex("tts_notification")) == 1,
            increaseVolume = cursor.getInt(cursor.getColumnIndex("increase_volume"))
    )

    val isRepeatedAlarm
        get() = days > 0

    val nextTimeWithTicks: Long
        get() = nextTime + TimeUnit.MINUTES.toMillis(ticksTime.toLong())

    private val nextTimeFormatted: String
        get() {
            val format = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US)
            return format.format(Date(nextTime))
        }

    fun updateNextAlarmDate(manually: Boolean) {
        val calendar = Calendar.getInstance()
        val calendarNow = Calendar.getInstance()
        nextTime = -1L
        if (exactDate == 0L) {
            calendar[Calendar.HOUR_OF_DAY] = hour
            calendar[Calendar.MINUTE] = minute
            calendar[Calendar.SECOND] = 0
            calendar[Calendar.MILLISECOND] = 0
            if (!manually) {
                calendarNow.add(Calendar.MINUTE, ticksTime)
            }
            if (calendar.before(calendarNow)) {
                calendar.add(Calendar.DAY_OF_YEAR, 1)
            }
            if (days > 0) {
                val allDaysAsList = allDaysAsList()
                var i = calendar[Calendar.DAY_OF_WEEK] - 1
                var skip = canceledNextAlarms
                while (true) {
                    if (i >= allDaysAsList.size) {
                        i = 0
                    }
                    val fireAtThisDayOfWeek = allDaysAsList[i]
                    if (fireAtThisDayOfWeek) {
                        if (skip <= 0) {
                            break
                        } else {
                            skip--
                            if (nextTime == -1L) {
                                nextTime = calendar.timeInMillis
                            }
                        }
                    }
                    calendar.add(Calendar.DAY_OF_YEAR, 1)
                    i++
                }
            }
        } else {
            calendar.timeInMillis = exactDate
            if (calendar.before(calendarNow)) {
                calendar.add(Calendar.DAY_OF_YEAR, 1)
                exactDate = calendar.timeInMillis
            }
        }

        nextNotCanceledTime = calendar.timeInMillis
        if (nextTime == -1L) {
            if (ticksTime > 0) {
                calendar.add(Calendar.MINUTE, -ticksTime)
            }
            nextTime = calendar.timeInMillis
        }
        nextRequestCode = alarmRequestCode
    }

    val alarmRequestCode: Int
        get() = id.toInt() * 100000

    val upcomingAlarmRequestCode: Int
        get() = alarmRequestCode + 1

    val snoozedAlarmRequestCode: Int
        get() = alarmRequestCode + 2

    val fullScreenIntentCode: Int
        get() = alarmRequestCode + 3


    private fun allDaysAsList(): List<Boolean> {
        return listOf(
                days and SU_BINARY == SU_BINARY,
                days and MO_BINARY == MO_BINARY,
                days and TU_BINARY == TU_BINARY,
                days and WE_BINARY == WE_BINARY,
                days and TH_BINARY == TH_BINARY,
                days and FR_BINARY == FR_BINARY,
                days and SA_BINARY == SA_BINARY)
    }

    override fun toString(): String {
        return "AlarmEntity(id=$id, hour=$hour, minute=$minute, nextTime=$nextTimeFormatted, " +
                "exactDate=$exactDate, ticksTime=$ticksTime, melodyUrl=$melodyUrl, canceled=$canceledNextAlarms, " +
                "complexity=$complexity, message=$message, isActive=$isActive, days=$days)"
    }

    val daysAsMap: Map<Int, Boolean>
        get() {
            val daysMap: MutableMap<Int, Boolean> = LinkedHashMap()
            daysMap[R.string.mo] = days and MO_BINARY == MO_BINARY
            daysMap[R.string.tu] = days and TU_BINARY == TU_BINARY
            daysMap[R.string.we] = days and WE_BINARY == WE_BINARY
            daysMap[R.string.th] = days and TH_BINARY == TH_BINARY
            daysMap[R.string.fr] = days and FR_BINARY == FR_BINARY
            daysMap[R.string.sa] = days and SA_BINARY == SA_BINARY
            daysMap[R.string.su] = days and SU_BINARY == SU_BINARY
            return daysMap
        }

    companion object {
        const val MO_BINARY = 128
        const val TU_BINARY = 64
        const val WE_BINARY = 32
        const val TH_BINARY = 16
        const val SA_BINARY = 4
        const val FR_BINARY = 8
        const val SU_BINARY = 2
        fun daysMarshaling(mo: Boolean, tu: Boolean, we: Boolean, th: Boolean, fr: Boolean, sa: Boolean, su: Boolean): Int {
            var d = 0
            if (mo) {
                d = d or MO_BINARY
            }
            if (tu) {
                d = d or TU_BINARY
            }
            if (we) {
                d = d or WE_BINARY
            }
            if (th) {
                d = d or TH_BINARY
            }
            if (fr) {
                d = d or FR_BINARY
            }
            if (sa) {
                d = d or SA_BINARY
            }
            if (su) {
                d = d or SU_BINARY
            }
            return d
        }
    }
}